causync.py

```diff
@@ -0,0 +1,374 @@
+
+    def get_lockfile_path(self):
+        """ Returns the lockfile path, depends on the source directory. """
+
+        src_parent_dir = self.get_parent_dir(self.src_abs)
+        src_basedir = self.get_basename(self.src_abs)
+        src_basedir = "{}.lock".format(src_basedir)
+        return os.path.join(src_parent_dir, src_basedir)
+
+    def get_dirdate(self, dirname):
+        """ Returns the date extracted from a backup directory name.
+            Example: 'causync_180410_111237' results in a datetime object for '18-04-10 11:12:37'
+        """
+
+        try:
+            if isinstance(dirname, datetime):
+                return dirname
+            dirdate = datetime.strptime(dirname, self.config.DATE_FORMAT)
+            return dirdate
+        except ValueError as e:
+            self.logger.error(e)
+            exit(1)
+
+    def find_latest_backups(self, dirnames, count=5):
+        """ Returns the latest daily backup directory names. """
+
+        if len(dirnames) == 0:
+            return []
+
+        # extract dates from directory names and sort them
+        dirdates = [ self.get_dirdate(d) for d in dirnames ]
+        dirdates.sort(reverse=True)
+
+        # determine list length (if len < delete count, we don't do anything)
+        list_len = count if len(dirdates) >= count else len(dirdates)
+        # get a reverse sorted list of dirdates (descending by date)
+        latest_dates = dirdates[:list_len]
+        # convert them to string for rsync --link-dest (example: datetime obj becomes 'YYMMHH' string)
+        latest_dates = [ i.strftime(config.DATE_FORMAT) for i in latest_dates ]
+        # join each one with the destination directory (example: '/path/dest/sourcedir_YYMMHH'
+        latest_dates = [ os.path.join(self.dst_abs, i) for i in latest_dates ]
+
+        return latest_dates
+
+    def find_old_backups(self, dirnames, ival='daily', count=5):
+        """ Returns old backups we should delete.
+            The time interval is specified by 'ival'. Examples: daily, weekly, monthly, yearly.
+        """
+
+        curdate = datetime.now()
+        multiplier = timedelta(days=self.config.BACKUP_MULTIPLIERS[ival])
+
+        # extract dates from directory names and sort them
+        dirdates = list(sorted([ self.get_dirdate(d) for d in dirnames ]))
+
+        (keep, delete) = (list(), list())
+
+        for d in dirdates:
+            m = multiplier * count + multiplier
+            keepdate = curdate - m
+
+            if ival == 'yearly' and d.day != 1 and d.month != 1:
+                continue
+            elif ival == 'monthly' and d.day != 1:
+                continue
+            elif ival == 'weekly' and d.weekday() != 0:
+                continue
+
+            if d > keepdate:
+                keep.append(d)
+            else:
+                delete.append(d)
+
+        delete.sort(reverse=True)
+        dirdates.sort(reverse=True)
+        keep.sort(reverse=True)
+
+        return keep, delete
```

Comment:
What is the purpose of calculating and returning keep?
You do not seem to use it anywhere...